modules/3d/src/spc.cpp

```diff
@@ -0,0 +1,244 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Author: Zhangjie Chen <zhangjiec01@gmail.com>
+
+#include "opencv2/core.hpp"
+#include "opencv2/3d.hpp"
+#include "queue"
+#include "unordered_map"
+
+namespace cv{
+SpectralCluster::SpectralCluster() {
+    pImpl = makePtr<Impl>(0.1f, 0.05f);
+}
+
+SpectralCluster::SpectralCluster(float delta_val, float eta_val) {
+    CV_Assert(delta_val >= 0.f && delta_val <= 1.f);
+    CV_Assert(eta_val >= 0.f && eta_val <= 1.f);
+    pImpl = makePtr<Impl>(delta_val, eta_val);
+}
+
+class SpectralCluster::Impl {
+public:
+    float delta;
+    float eta;
+
+    Impl(float delta_val, float eta_val) : delta(delta_val), eta(eta_val) {}
+    ~Impl() = default;
+
+    static void getLaplacianMat(cv::Mat& in, cv::Mat& out);
+
+    void getAdjacentDistanceMat(cv::Mat& out, const std::vector<cv::Point3f>& vertices,
+                                const std::vector<std::vector<int32_t>>& indices) const;
+
+    static void getAffinityMat(cv::Mat& in, cv::Mat& out, const std::vector<std::vector<int32_t>>& indices);
+
+    static float getGeodesicDistance(const std::vector<cv::Point3f>& face1, const std::vector<cv::Point3f>& face2,
+                                     const std::pair<cv::Point3f, cv::Point3f>& edge);
+
+    float getAngleDistance(const std::vector<cv::Point3f>& face1, const std::vector<cv::Point3f>& face2) const;
+
+    static cv::Point3f calculateFaceCentroid(const std::vector<cv::Point3f>& face);
+
+    static cv::Point3f calculateFaceNormal(const std::vector<cv::Point3f>& face);
+};
+
+void SpectralCluster::Impl::getLaplacianMat(Mat &in, Mat &out) {
+    Mat degree_mat;
+    reduce(in, degree_mat, 1, REDUCE_SUM);
+    degree_mat = 1.0f / degree_mat;
+    sqrt(degree_mat, degree_mat);
+    Mat degree_mat_sqrt;
+    repeat(degree_mat, 1, in.rows, degree_mat_sqrt);
+    out = ((in.mul(degree_mat_sqrt).t()).mul(degree_mat_sqrt)).t();
```

Comment:
Do we really need these two consequent transpositions? Looks like this can be shortened.